lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,586 @@
+        cloneAndSetup(
+            newMaterialID, newTextureID, newImageID,
+            materialID, textureID, imageID,
+            allMaterials, allTextures, allImages);
+        primitive.material = newMaterialID;
+    } else if(texturesSeen.hasOwnProperty(textureID)) {
+        // Check if the texture is unique. If not clone the texture and the image.
+        cloneAndSetup(
+            undefined, newTextureID, newImageID,
+            materialID, textureID, imageID,
+            allMaterials, allTextures, allImages);
+        values.diffuse = newTextureID;
+    } else if(imagesSeen.hasOwnProperty(imageID)) {
+        // Check if the image is unique. if not, clone the image.
+        var texture = allTextures[textureID];
+        cloneAndSetup(
+            undefined, undefined, newImageID,
+            materialID, textureID, imageID,
+            allMaterials, allTextures, allImages);
+        texture.source = newImageID;
+    } else {
+        // If nothing was cloned, mark this material, texture, and image as seen
+        materialsSeen[materialID] = true;
+        texturesSeen[textureID] = true;
+        imagesSeen[imageID] = true;
+        newImageID = imageID;
+    }
+    return allImages[newImageID];
+}
+
+////////// loading //////////
+
+function generateRaytracerScene(gltf, scene, options) {
+    // Set up data we need for sampling. generate "triangle soup" over the whole scene.
+    var accessors = gltf.accessors;
+
+    // read all accessors in one go to avoid repeated read-and-conversion
+    var bufferDataByAccessor = {};
+
+    for (var accessorID in accessors) {
+        if (accessors.hasOwnProperty(accessorID)) {
+            var accessor = accessors[accessorID];
+            bufferDataByAccessor[accessorID] = readAccessor(gltf, accessor);
+        }
+    }
+
+    var raytracerScene = {
+        bufferDataByAccessor: bufferDataByAccessor,
+        numberSamples: defaultValue(options.numberSamples, 16),
+        rayDepth: defaultValue(options.rayDepth, 1.0), // TODO: compute dynamic default ray depth?
+        triangleSoup: [],
+        aoBufferByPrimitive: {},
+        nearCull: 0.5 / options.resolution
+    };
+
+    // TODO: currently assuming each primitive appears in the scene once. figure out what to do when this is not true.
+
+    // Generate all the world transform matrices
+    NodeHelpers.computeFlatTransformScene(scene, gltf.nodes);
+
+    var parameters = {
+        resolution: options.resolution,
+        raytracerScene: raytracerScene
+    };
+
+    NodeHelpers.forEachPrimitiveInScene(gltf, scene, processPrimitive, parameters);
+
+    return raytracerScene;
+}
+
+function processPrimitive(primitive, meshPrimitiveID, parameters, node) {
+    // AO only works with triangles
+    if (defined(primitive.mode) && primitive.mode !== WebGLConstants.TRIANGLES) {
+        throw new DeveloperError('In stage bakeAmbientOcclusion: baking AO only supports triangle primitives.');
```

Comment:
I don't think we should throw here.  Some models will have polylines and it would be fine to just skip them.

@lilleyse as part of the cleanup roadmap, can you add an item for logging, e.g., there might be a callback function that could be used to log for the console (for the command-line tool) and return a message to a client (for the web service).